File: C++/SearchForARange.cc
// Given a sorted array of integers, find the starting and ending position of a given target value.

// Your algorithm's runtime complexity must be in the order of O(log n).

// If the target is not found in the array, return [-1, -1].

// For example,
// Given [5, 7, 7, 8, 8, 10] and target value 8,
// return [3, 4]. 

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {        
        auto lowIt=std::lower_bound (nums.begin(), nums.end(), target);
        auto upIt= std::upper_bound (nums.begin(), nums.end(), target);
        int low = *lowIt == target ? lowIt - nums.begin() : -1;
        int up = *(upIt - 1) == target ? upIt - nums.begin() - 1 : -1;
        return {low, up};
    }
};

File: Java/SearchForARange.java
// Given a sorted array of integers, find the starting and ending position of a given target value.

// Your algorithm's runtime complexity must be in the order of O(log n).

// If the target is not found in the array, return [-1, -1].

// For example,
// Given [5, 7, 7, 8, 8, 10] and target value 8,
// return [3, 4].

public class SearchForARange {
    public int[] searchRange(int[] nums, int target) {
        return new int[]{lowerBound(nums, target), upperBound(nums, target)};
    }
    
    int lowerBound(int[] nums, int target) {
        int l = 0, r = nums.length - 1;
        while (l <= r) {
            int m = l + (r - l) / 2;
            if (nums[m] == target && (m == 0 || nums[m-1] != nums[m])) {
                return m;
            } else if (nums[m] < target) {
                l = m + 1;
            } else {
                r = m - 1;
            }            
        }
        return -1;
    }
    
    int upperBound(int[] nums, int target) {
        int l = 0, r = nums.length - 1;
        while (l <= r) {
            int m = l + (r - l) / 2;
            if (nums[m] == target && (m == nums.length - 1 || nums[m+1] != nums[m])) {
                return m;
            } else if (nums[m] <= target) {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }
        return -1;
    }
}
